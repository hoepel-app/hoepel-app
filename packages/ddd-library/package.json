{
  "name": "@hoepel.app/ddd-library",
  "version": "0.3.113",
  "main": "dist/index.js",
  "license": "GPL-3.0-or-later",
  "devDependencies": {
    "typescript": "^4.1.3"
  },
  "scripts": {
    "build": "tsc"
  },
  "publishConfig": {
    "access": "public"
  }
}
