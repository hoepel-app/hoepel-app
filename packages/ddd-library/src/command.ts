import { Message } from './message'

export type CommandRequestedBy = {
  type: 'user'
  uid: string
  email: string
}

export type CommandRejected = { status: 'rejected' }
export type CommandAccepted = { status: 'accepted' }
export type CommandResult = CommandAccepted | CommandRejected

export type Command<T> = {
  /** @deprecated */
  toProps(): T

  requestedBy: CommandRequestedBy
} & Message<T>
