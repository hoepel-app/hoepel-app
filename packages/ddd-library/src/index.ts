export * from './command'
export * from './event'
export * from './id-generator'
export * from './aggregate'
export * from './command-base'
