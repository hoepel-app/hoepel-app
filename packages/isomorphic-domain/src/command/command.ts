export type CommandRejected = { status: 'rejected' }
export type CommandAccepted = { status: 'accepted' }
export type CommandResult = CommandAccepted | CommandRejected

export type Command<T> = {
  toProps(): T
}
