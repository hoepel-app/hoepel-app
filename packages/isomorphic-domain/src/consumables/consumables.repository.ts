import { Observable } from 'rxjs'
import { Consumables } from './consumables'

export type ConsumablesRepository = {
  getForTenant(tenantId: string): Observable<Consumables>
  put(entity: Consumables): Promise<void>
}
