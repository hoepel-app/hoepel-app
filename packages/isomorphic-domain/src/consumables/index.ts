export { AddConsumableCommand } from './commands/add-consumable.command'
export { ChangeConsumablePriceCommand } from './commands/change-consumable-price.command'
export { RemoveConsumableCommand } from './commands/remove-consumable.command'
export { ConsumablesApplicationService } from './consumables.application-service'
export { ConsumablesRepository } from './consumables.repository'
export {
  Consumables,
  ConsumablesProps,
  Consumable,
  ConsumableProps,
} from './consumables'
