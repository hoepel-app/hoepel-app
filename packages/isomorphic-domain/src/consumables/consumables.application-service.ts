import { ConsumablesRepository } from './consumables.repository'
import { Observable } from 'rxjs'
import { Consumables } from './consumables'
import { AddConsumableCommand } from './commands/add-consumable.command'
import { first } from 'rxjs/operators'
import { CommandResult } from '@hoepel.app/ddd-library/src'
import { ChangeConsumablePriceCommand } from './commands/change-consumable-price.command'
import { RemoveConsumableCommand } from './commands/remove-consumable.command'
import { RenameConsumableCommand } from './commands/rename-consumable.command'

export class ConsumablesApplicationService {
  constructor(private readonly repo: ConsumablesRepository) {}

  findConsumables(tenantId: string): Observable<Consumables> {
    return this.repo.getForTenant(tenantId)
  }

  async addConsumable(command: AddConsumableCommand): Promise<CommandResult> {
    const consumables = await this.findConsumables(command.tenantId)
      .pipe(first())
      .toPromise()

    if (!consumables.mayAddConsumable(command.consumable)) {
      return { status: 'rejected' }
    }

    await this.repo.put(consumables.withConsumableAdded(command.consumable))

    return { status: 'accepted' }
  }

  async changeConsumablePrice(
    command: ChangeConsumablePriceCommand
  ): Promise<CommandResult> {
    const consumables = await this.findConsumables(command.tenantId)
      .pipe(first())
      .toPromise()

    if (
      command.price.totalCents < 0 ||
      !consumables.consumableWithNameExists(command.consumableName)
    ) {
      return { status: 'rejected' }
    }

    await this.repo.put(
      consumables.withChangedPriceForConsumable(
        command.consumableName,
        command.price.totalCents
      )
    )

    return { status: 'accepted' }
  }

  async renameConsumable(
    command: RenameConsumableCommand
  ): Promise<CommandResult> {
    const consumables = await this.findConsumables(command.tenantId)
      .pipe(first())
      .toPromise()

    if (consumables.consumableWithNameExists(command.newName)) {
      return { status: 'rejected', reason: 'Consumable with new name exists' }
    }

    if (!consumables.consumableWithNameExists(command.oldName)) {
      return { status: 'rejected', reason: 'Consumable not found' }
    }

    await this.repo.put(
      consumables.withRenamedConsumable(command.oldName, command.newName)
    )

    return { status: 'accepted' }
  }

  async removeConsumable(
    command: RemoveConsumableCommand
  ): Promise<CommandResult> {
    const consumables = await this.findConsumables(command.tenantId)
      .pipe(first())
      .toPromise()

    if (!consumables.consumableWithNameExists(command.consumableName)) {
      return { status: 'rejected' }
    }

    await this.repo.put(
      consumables.withConsumableRemoved(command.consumableName)
    )

    return { status: 'accepted' }
  }
}
