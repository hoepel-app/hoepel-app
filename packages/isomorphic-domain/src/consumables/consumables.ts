import { Price } from '@hoepel.app/types'
import { Aggregate } from '@hoepel.app/ddd-library/src'

export type ConsumableProps = {
  readonly name: string
  readonly priceCents: number
}

export type ConsumablesProps = {
  readonly consumables: readonly ConsumableProps[]
  readonly tenantId: string
}

export class Consumables implements Aggregate {
  private constructor(private readonly props: ConsumablesProps) {}

  static createEmpty(tenantId: string): Consumables {
    return new Consumables({ tenantId, consumables: [] })
  }

  static fromProps(props: ConsumablesProps): Consumables {
    return new Consumables(props)
  }

  toProps(): ConsumablesProps {
    return this.props
  }

  get id(): string {
    return `${this.tenantId}-consumables`
  }

  get tenantId(): string {
    return this.props.tenantId
  }

  get names(): ReadonlySet<string> {
    return new Set(this.consumables.map((consumable) => consumable.name))
  }

  get consumables(): readonly Consumable[] {
    return this.props.consumables.map((props) => Consumable.fromProps(props))
  }

  withConsumableAdded(consumable: Consumable): Consumables {
    if (!this.mayAddConsumable(consumable)) {
      return this
    }

    return Consumables.fromProps({
      ...this.toProps(),
      consumables: [...this.toProps().consumables, consumable.toProps()],
    })
  }

  withConsumableRemoved(name: string): Consumables {
    return Consumables.fromProps({
      ...this.toProps(),
      consumables: this.toProps().consumables.filter((c) => c.name !== name),
    })
  }

  withChangedPriceForConsumable(
    name: string,
    newPriceCents: number
  ): Consumables {
    if (!this.consumableWithNameExists(name)) {
      return this
    }

    return Consumables.fromProps({
      ...this.toProps(),
      consumables: [
        ...this.toProps().consumables.filter((c) => c.name !== name),
        Consumable.create(name, newPriceCents).toProps(),
      ],
    })
  }

  withRenamedConsumable(oldName: string, newName: string): Consumables {
    const currentConsumable = this.findConsumableByName(oldName)

    if (this.consumableWithNameExists(newName) || currentConsumable == null) {
      return this
    }

    return Consumables.fromProps({
      ...this.toProps(),
      consumables: [
        ...this.toProps().consumables.filter((c) => c.name !== oldName),
        currentConsumable.withName(newName).toProps(),
      ],
    })
  }

  findConsumableByName(name: string): Consumable | null {
    return (
      this.consumables.find((consumable) => consumable.name === name) ?? null
    )
  }

  mayAddConsumable(consumable: Consumable): boolean {
    return !this.consumableWithNameExists(consumable.name)
  }

  consumableWithNameExists(name: string): boolean {
    return this.findConsumableByName(name) != null
  }
}

/**
 * Used to represent something that can be consumed or purchased, which is not an attendance.
 * E.g. drink, cookies, ...
 * Consumption of consumables is tied to a dayid/DayDate o(a specific date)
 * The same consumable can be consumed multiple times by the same child on the same day.
 */
export class Consumable {
  private constructor(private readonly props: ConsumableProps) {}

  static create(name: string, priceCents: number): Consumable {
    if (priceCents < 0) {
      throw new Error('Price in cents must be a positive number')
    }

    if (Math.round(priceCents) != priceCents) {
      throw new Error('Price in cents must be an integer')
    }

    return new Consumable({ name, priceCents })
  }

  static fromProps(props: ConsumableProps): Consumable {
    return new Consumable(props)
  }

  withName(newName: string): Consumable {
    return Consumable.fromProps({
      ...this.toProps(),
      name: newName,
    })
  }

  toProps(): ConsumableProps {
    return this.props
  }

  /**
   * Price of the consumable. May not be included in fiscal certificates (these costs are non-refundable)!
   */
  get price(): Price {
    return Price.fromCents(this.props.priceCents)
  }

  /**
   * Name of the consumable, e.g. "Big chocolate cookie"
   */
  get name(): string {
    return this.props.name
  }
}
