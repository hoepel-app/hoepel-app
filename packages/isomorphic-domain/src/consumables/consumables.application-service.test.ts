import { Consumables, Consumable } from './consumables'
import { ConsumablesApplicationService } from './consumables.application-service'
import { CommandMetadata } from '@hoepel.app/ddd-library'
import { of } from 'rxjs'
import { first } from 'rxjs/operators'
import { AddConsumableCommand } from './commands/add-consumable.command'
import { ChangeConsumablePriceCommand } from './commands/change-consumable-price.command'
import { RemoveConsumableCommand } from './commands/remove-consumable.command'

describe('ConsumablesApplicationService', () => {
  const exampleConsumables = (tenantId: string): Consumables =>
    Consumables.createEmpty(tenantId)
      .withConsumableAdded(Consumable.create('Big Chocolate Cookie', 100))
      .withConsumableAdded(Consumable.create('Water', 0))

  const commandMetadata: CommandMetadata = {
    commandId: 'my-command-id-123',
    requestedBy: {
      email: 'test@example.org',
      type: 'user',
      uid: 'my-uid-123',
    },
    tenantId: 'some-tenant-id',
    timestamp: new Date('2020-04-25T19:43:29.161Z'),
  }

  describe('findConsumables', () => {
    it('finds consumable for a tenant', async () => {
      const repo = {
        getForTenant: jest.fn((tenantId: string) =>
          of(exampleConsumables(tenantId))
        ),
        put: jest.fn(),
      }

      const service = new ConsumablesApplicationService(repo)

      const found = await service
        .findConsumables('my-tenant-id-here')
        .pipe(first())
        .toPromise()

      expect(repo.getForTenant).toHaveBeenCalledTimes(1)
      expect(repo.getForTenant).toHaveBeenLastCalledWith('my-tenant-id-here')
      expect(found).toEqual(exampleConsumables('my-tenant-id-here'))
    })
  })

  describe('addConsumable', () => {
    it('rejects when adding duplicate consumable', async () => {
      const repo = {
        getForTenant: jest.fn((tenantId: string) =>
          of(exampleConsumables(tenantId))
        ),
        put: jest.fn(),
      }

      const service = new ConsumablesApplicationService(repo)
      const newConsumable = Consumable.create('Water', 200)
      const command = AddConsumableCommand.create(
        newConsumable,
        commandMetadata
      )

      const commandResult = await service.addConsumable(command)

      expect(commandResult.status).toEqual('rejected')
      expect(repo.put).not.toHaveBeenCalled()
    })

    it('adds a new consumable', async () => {
      const repo = {
        getForTenant: jest.fn((tenantId: string) =>
          of(exampleConsumables(tenantId))
        ),
        put: jest.fn(),
      }

      const service = new ConsumablesApplicationService(repo)
      const newConsumable = Consumable.create('Soup', 100)
      const command = AddConsumableCommand.create(
        newConsumable,
        commandMetadata
      )

      const commandResult = await service.addConsumable(command)

      expect(commandResult.status).toEqual('accepted')
      expect(repo.put).toHaveBeenCalledTimes(1)
      expect(repo.put.mock.calls[0]).toMatchSnapshot()
    })
  })

  describe('changeConsumablePrice', () => {
    it('rejects when consumable does not exist', async () => {
      const repo = {
        getForTenant: jest.fn((tenantId: string) =>
          of(exampleConsumables(tenantId))
        ),
        put: jest.fn(),
      }

      const service = new ConsumablesApplicationService(repo)
      const command = ChangeConsumablePriceCommand.create(
        'Lemonade',
        100,
        commandMetadata
      )

      const commandResult = await service.changeConsumablePrice(command)

      expect(commandResult.status).toEqual('rejected')
      expect(repo.put).not.toHaveBeenCalled()
    })

    it('changes the price for a consumable', async () => {
      const repo = {
        getForTenant: jest.fn((tenantId: string) =>
          of(exampleConsumables(tenantId))
        ),
        put: jest.fn(),
      }

      const service = new ConsumablesApplicationService(repo)
      const command = ChangeConsumablePriceCommand.create(
        'Big Chocolate Cookie',
        50,
        commandMetadata
      )

      const commandResult = await service.changeConsumablePrice(command)

      expect(commandResult.status).toEqual('accepted')
      expect(repo.put).toHaveBeenCalledTimes(1)
      expect(repo.put.mock.calls[0]).toMatchSnapshot()
    })
  })

  describe('removeConsumable', () => {
    it('rejects when consumable does not exist', async () => {
      const repo = {
        getForTenant: jest.fn((tenantId: string) =>
          of(exampleConsumables(tenantId))
        ),
        put: jest.fn(),
      }

      const service = new ConsumablesApplicationService(repo)
      const command = RemoveConsumableCommand.create(
        'Lemonade',
        commandMetadata
      )

      const commandResult = await service.removeConsumable(command)

      expect(commandResult.status).toEqual('rejected')
      expect(repo.put).not.toHaveBeenCalled()
    })

    it('removes a consumable', async () => {
      const repo = {
        getForTenant: jest.fn((tenantId: string) =>
          of(exampleConsumables(tenantId))
        ),
        put: jest.fn(),
      }

      const service = new ConsumablesApplicationService(repo)
      const command = RemoveConsumableCommand.create(
        'Big Chocolate Cookie',
        commandMetadata
      )

      const commandResult = await service.removeConsumable(command)

      expect(commandResult.status).toEqual('accepted')
      expect(repo.put).toHaveBeenCalledTimes(1)
      expect(repo.put.mock.calls[0]).toMatchSnapshot()
    })
  })
})
