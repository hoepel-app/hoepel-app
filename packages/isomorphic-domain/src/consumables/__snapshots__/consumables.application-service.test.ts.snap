// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`ConsumablesApplicationService addConsumable adds a new consumable 1`] = `
Array [
  Consumables {
    "props": Object {
      "consumables": Array [
        Object {
          "name": "Big Chocolate Cookie",
          "priceCents": 100,
        },
        Object {
          "name": "Water",
          "priceCents": 0,
        },
        Object {
          "name": "Soup",
          "priceCents": 100,
        },
      ],
      "tenantId": "some-tenant-id",
    },
  },
]
`;

exports[`ConsumablesApplicationService changeConsumablePrice changes the price for a consumable 1`] = `
Array [
  Consumables {
    "props": Object {
      "consumables": Array [
        Object {
          "name": "Water",
          "priceCents": 0,
        },
        Object {
          "name": "Big Chocolate Cookie",
          "priceCents": 50,
        },
      ],
      "tenantId": "some-tenant-id",
    },
  },
]
`;

exports[`ConsumablesApplicationService removeConsumable removes a consumable 1`] = `
Array [
  Consumables {
    "props": Object {
      "consumables": Array [
        Object {
          "name": "Water",
          "priceCents": 0,
        },
      ],
      "tenantId": "some-tenant-id",
    },
  },
]
`;

exports[`ConsumablesApplicationService renameConsumable renames consumable 1`] = `
Array [
  Consumables {
    "props": Object {
      "consumables": Array [
        Object {
          "name": "Water",
          "priceCents": 0,
        },
        Object {
          "name": "Cookie",
          "priceCents": 100,
        },
      ],
      "tenantId": "some-tenant-id",
    },
  },
]
`;
